packer/plugin-getter/plugins.go

```diff
@@ -0,0 +1,421 @@
+	// safety but can also be relative.
+	FromFolders []string
+
+	BinaryInstallationOptions
+}
+
+func (pr Requirement) filenamePrefix() string {
+	return "packer-plugin-" + pr.Identifier.Type + "_"
+}
+
+func (opts BinaryInstallationOptions) filenameSuffix() string {
+	return "_" + opts.OS + "_" + opts.ARCH + opts.Extension
+}
+
+// ListInstallations lists unique installed versions of plugin Requirement pr
+// with opts as a filter.
+//
+// Installations are sorted by version and one binary per version is returned.
+// Last binary detected takes precedence: in the order 'FromFolders' option.
+//
+// You must pass at least one option to Checksumers for a binary to be even
+// consider.
+func (pr Requirement) ListInstallations(opts ListInstallationsOptions) (InstallList, error) {
+	res := InstallList{}
+	filenamePrefix := pr.filenamePrefix()
+	filenameSuffix := opts.filenameSuffix()
+	log.Printf("[TRACE] listing potential installations for %q that match %q. %#v", pr.Identifier.ForDisplay(), pr.VersionConstraints, opts)
+	for _, knownFolder := range opts.FromFolders {
+		glob := filepath.Join(knownFolder, pr.Identifier.Hostname, pr.Identifier.Namespace, pr.Identifier.Type, filenamePrefix+"*"+filenameSuffix)
+
+		matches, err := filepath.Glob(glob)
+		if err != nil {
+			return nil, fmt.Errorf("ListInstallations: %q failed to list binaries in folder: %v", pr.Identifier.String(), err)
+		}
+		for _, path := range matches {
+			fname := filepath.Base(path)
+			if fname == "." {
+				continue
+			}
+
+			// base name could look like packer-plugin-amazon_v1.2.3_darwin_amd64_x4
+			versionStr := strings.TrimPrefix(fname, filenamePrefix)
+			versionStr = strings.TrimSuffix(versionStr, filenameSuffix)
+			pv, err := version.NewVersion(versionStr)
+			if err != nil {
+				// could not be parsed, ignoring the file
+				log.Printf("found %q with an incorrect %q version, ignoring it. %v", path, versionStr, err)
+				continue
+			}
+
+			// no constraint means always pass, this will happen for implicit
+			// plugin requirements
+			if !pr.VersionConstraints.Check(pv) {
+				log.Printf("[TRACE] version %q of file %q does not match constraint %q", versionStr, path, pr.VersionConstraints.String())
+				continue
+			}
+
+			checksumOk := false
+			for _, checksummer := range opts.Checksummers {
+
+				cs, err := checksummer.GetChecksumOfFile(path)
+				if err != nil {
+					log.Printf("[TRACE] GetChecksumOfFile(%q) failed: %v", path, err)
+					continue
+				}
+
+				if err := checksummer.ChecksumFile(cs, path); err != nil {
+					log.Printf("[TRACE] ChecksumFile(%q) failed: %v", path, err)
+					continue
+				}
+				checksumOk = true
+				break
+			}
+			if !checksumOk {
+				log.Printf("[TRACE] No checksum found for %q ignoring possibly unsafe binary", path)
+				continue
+			}
+
+			res.InsertSortedUniq(&Installation{
+				BinaryPath: path,
+				Version:    versionStr,
+			})
+		}
+	}
+	return res, nil
+}
+
+// InstallList is a list of installs
```

Comment:
this could use clarification: list of things already installed, or list of things that we need to install?

Reply:
Yup !